tests/integration/config_persistence_test.py

```diff
@@ -0,0 +1,222 @@
+        version=1,
+    )
+    source_names = ['monitor1', 'monitor2']
+
+    # Create params with custom values (non-default)
+    custom_params = MonitorDataParams(
+        toa_edges=TOAEdges(
+            start=5.0,
+            stop=15.0,
+            num_bins=150,
+            scale=Scale.LINEAR,
+            unit=TimeUnit.MS,
+        )
+    )
+
+    # Start the workflow with custom parameters
+    job_ids = backend_with_null_transport.workflow_controller.start_workflow(
+        workflow_id=workflow_id,
+        source_names=source_names,
+        config=custom_params,
+    )
+
+    assert len(job_ids) == 2, f"Expected 2 jobs, got {len(job_ids)}"
+
+    # Verify params are stored in config store
+    stored_config = backend_with_null_transport.workflow_controller.get_workflow_config(
+        workflow_id
+    )
+    assert stored_config is not None, "Config should be stored in config store"
+    assert stored_config.source_names == source_names
+    assert stored_config.params == custom_params.model_dump()
+
+    # Create adapter and verify it retrieves correct params from config store
+    adapter = backend_with_null_transport.workflow_controller.create_workflow_adapter(
+        workflow_id
+    )
+    assert adapter.initial_source_names == source_names
+    assert adapter.initial_parameter_values == custom_params.model_dump()
+
+
+def test_adapter_filters_removed_sources(
+    backend_with_null_transport: DashboardBackend,
+) -> None:
+    """
+    Test that adapter filters out sources that are no longer available.
+
+    This test verifies that if a workflow was started with sources that are
+    no longer in the workflow spec, the adapter correctly filters them out
+    when restoring the configuration.
+    """
+    workflow_id = WorkflowId(
+        instrument='dummy',
+        namespace='monitor_data',
+        name='monitor_histogram',
+        version=1,
+    )
+
+    # Start workflow with multiple sources (monitor3 not in spec)
+    source_names = ['monitor1', 'monitor2', 'monitor3']
+    backend_with_null_transport.workflow_controller.start_workflow(
+        workflow_id=workflow_id,
+        source_names=source_names,
+        config=MonitorDataParams(),
+    )
+
+    # Verify config is stored with all sources
+    stored_config = backend_with_null_transport.workflow_controller.get_workflow_config(
+        workflow_id
+    )
+    assert stored_config is not None
+    assert set(stored_config.source_names) == {'monitor1', 'monitor2', 'monitor3'}
+
+    # Create adapter - it should filter to only sources available in spec
+    adapter = backend_with_null_transport.workflow_controller.create_workflow_adapter(
+        workflow_id
+    )
+
+    # Adapter should only return sources that are both persisted AND in the spec
+    initial_sources = adapter.initial_source_names
+    assert 'monitor1' in initial_sources
+    assert 'monitor2' in initial_sources
+    # monitor3 should be filtered out as it's not in the workflow spec
```

Comment:
Can you explain where in the test monitor3 gets dropped? I don't see it.

Also, we should probably test that it really does get dropped, something like
`assert 'monitor3' not in initial_sources`

Reply:
It's still not explicit. I see `set(stored_config.source_names) == {'monitor1', 'monitor2', 'monitor3'}`.
I can't figure out where it's happening.
Is it in the `backend_with_null_transport.workflow_controller.create_workflow_adapter`?